File: LinkedList/Doubly/linked_list.c
#include<stdio.h>
#include<stdlib.h>

typedef struct Node {
	int data;
	struct Node* next;
	struct Node* prev;
} Node;

int length = 0;

Node* create_node(int data){
	Node* new_node = malloc(sizeof(Node));
	new_node -> data = data;
	new_node->next = new_node->prev = NULL;
	return new_node;
}

void prepend(Node** head, Node** tail, int value){
	Node* new_node = create_node(value);
	length++;
	if(*head == NULL || *tail == NULL){
		*head = *tail = new_node;
		return;
	}

	Node* node = *head;
	new_node->next = node;
	node->prev = new_node;
	*head = new_node;
}

void append(Node** head, Node** tail, int value){
	Node* new_node = create_node(value);
	length++;
	if(*head == NULL || *tail == NULL){
		*head = *tail = new_node;
		return;
	}

	Node* node = *tail;
	node->next = new_node;
	new_node->prev = node;
	*tail = new_node;
}

void remove_head(Node** head, Node** tail){
	if(*head == NULL || *tail == NULL){
		printf("-=-=-=- Trying to remove an empty list -=-=-=-\n");
		return;
	}

	if(length == 1){
		*head = *tail = NULL;
		length--;
		return;
	}
	length--;	
	Node* node = *head;
	node->next->prev = NULL;
	*head = node->next;
	free(node);
}

void remove_tail(Node** head, Node** tail){
	if(*head == NULL || *tail == NULL){
		printf("-=-=-=- Trying to remove an empty list -=-=-=-\n");
		return;
	}

	if(length == 1){
		*head = *tail = NULL;
		length--;
		return;
	}
	length--;
	Node* node = *tail;
	node->prev->next = NULL;
	*tail = node->prev;
	free(node);
}

void traverse(Node* head){
	Node* node = head;
	printf(" --- Traversing Linked list ---\n");
	while(node != NULL){
		printf("%d -> ", node->data);
		node = node -> next;
	}
	printf(" NULL\n");
}

void traverse_reverse(Node* tail){
	Node* node = tail;
	printf(" --- Traversing Linked list ---\n");
	while(node != NULL){
		printf("%d -> ", node->data);
		node = node -> prev;
	}
	printf(" NULL\n");
}

void free_all(Node* head){
	Node* node = head;
	while(node != NULL){
		Node* node_to_delete = node;
		node = node -> next;
		free(node_to_delete);
	}
}

int main(){
	Node* head;
	Node* tail;
	head = tail = NULL;
	// 56 9 77 5 10 
	prepend(&head, &tail, 77);
	append(&head, &tail, 5);
	prepend(&head, &tail, 9);
	append(&head, &tail, 10);
	prepend(&head, &tail, 56);

	traverse(head);
	remove_head(&head, &tail);
	remove_head(&head, &tail);
	traverse_reverse(tail);

	free_all(head);
}

File: LinkedList/Doubly/linked_list.py
class Node:
    def __init__(self, data = None):
        self.data = data
        self.next = None
        self.prev = None

class DLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def prepend(self, value: int):
        new_node = Node(value)
        self.length += 1
        if self.head == None:
            self.head = new_node
            self.tail = new_node
            return
        new_node.next = self.head
        self.head.prev = new_node
        self.head = new_node

    def append(self, value: int):
        new_node = Node(value)
        self.length += 1
        if self.tail == None:
            self.head = new_node
            self.tail = new_node
            return

        self.tail.next = new_node
        new_node.prev = self.tail
        self.tail = new_node

    def remove_head(self):
        if self.head == None or self.tail == None:
            print("-=-=-=-=- Trying to remove an empty list -=-=-=-=-")
            return

        if self.length == 1:
            self.length -= 1
            self.head = self.tail = None
            return

        self.length -= 1
        node_to_delete = self.head
        self.head = node_to_delete.next
        self.head.prev = None
        node_to_delete = None

    def remove_tail(self):
        if self.head == None or self.tail == None:
            print("-=-=-=-=- Trying to remove an empty list -=-=-=-=-")
            return

        if self.length == 1:
            self.length -= 1
            self.head = self.tail = None
            return
        
        self.length -= 1
        node_to_delete = self.tail
        self.tail = node_to_delete.prev
        self.tail.next = None
        node_to_delete =  None


    def traverse(self):
        print("--- TRAVERSE LINKED LIST ---")
        node = self.head
        while node != None:
            print(f"{node.data} -> ", end = " ")
            node = node.next

        print("null")

    def traverse_reverse(self):
        print("--- TRAVERSE LINKED LIST IN REVERSE ---")
        node = self.tail
        while node != None:
            print(f"{node.data} -> ", end = " ")
            node = node.prev

        print("null")


# 56 9 77 5 10
node = DLinkedList()
node.prepend(77)
node.append(5)
node.prepend(9)
node.append(10)
node.prepend(56)

node.traverse()
node.remove_tail()
node.traverse()
node.remove_head()
node.traverse()
node.remove_tail()
node.remove_tail()
node.remove_tail()
node.remove_tail()
node.traverse()

